src/array.jl

```diff
@@ -972,3 +972,52 @@ Base.repeat(a::CatArrOrSub{T, N},
 Base.repeat(a::CatArrOrSub{T, N};
             inner = nothing, outer = nothing) where {T, N} =
     CategoricalArray{T, N}(repeat(refs(a), inner=inner, outer=outer), copy(pool(a)))
+
+# JSON3 writing/reading
+StructTypes.StructType(::Type{<:CategoricalVector{T}}) where T = StructTypes.ArrayType()
+StructTypes.construct(::Type{T}, x::Vector{S}; null_as_missing=false) where {T <: CategoricalVector,S} =
+    __categorical__(T,x,null_as_missing)
+StructTypes.construct(::Type{T}, x::Vector{S}; null_as_missing=false) where {T <: CategoricalArray,S} =
+    __categorical__(T,x,null_as_missing)
+
+# we need both CategoricalVector and CategoricalArray below; with the various type params
+# the one is not always <: of the other; users might reasonably request a type of
+# CategoricalArray rather than CategoricalVector
+__categorical__(::Type{<:CategoricalVector},array,asmissing) =
+    __categorical__general__(array,asmissing)
+__categorical__(::Type{<:CategoricalVector{Union{Missing,T}}},array,asmissing) where T =
+    __categorical__missing__(T,array)
+__categorical__(::Type{<:CategoricalVector{Union{Nothing,T}}},array,asmissing) where T =
+    __categorical__nothing__(T,array)
+__categorical__(::Type{<:CategoricalArray},array,asmissing) =
+    __categorical__general__(array,asmissing)
+__categorical__(::Type{<:CategoricalArray{Union{Missing,T}}},array,asmissing) where T =
+    __categorical__missing__(T,array)
+__categorical__(::Type{<:CategoricalArray{Union{Nothing,T}}},array,asmissing) where T =
+    __categorical__nothing__(T,array)
+
+function __categorical__general__(array,asmissing)
+    # check for small union with `Nothing` type
+    types = unique!(typeof.(array))
```

Comment:
Ah thanks! That's useful to know about.

The problem here is that `JSON3` does not handle small union types right now. It returns `Array{Any}` objects when there are numbers and `null` values, for example. So it needs to be a runtime check. @quinnj what're your thoughts on changing the the behavior of JSON3 to handle small union types?

Reply:
@haberdashPI, can you clarify what error you're using w/ small unions? It should work. For example:
```julia
julia> x = JSON3.read("[1.0, null, 2.0, 3.14]", Vector{Union{Missing, Float64}})
4-element Array{Union{Missing, Float64},1}:
 1.0
  missing
 2.0
 3.14
```

Reply:
Weird... okay, obviously it's my fault. It looks like there's a specific case where I am getting an eltype of `Any`. I'll dig into that and figure out what's going on.

Reply:
Oh! I see what's happening:`JSON3.read("[1.0, null, 2.0, 3.14]", CategoricalArray)` returns a type of `Any`, because it is using the `eltype` of the second argument.

Maybe what I'm trying to do with `__categorical_general__` isn't super useful. I had thought it might be handy to be able to call `JSON3.read("[1.0, null, 2.0, 3.14]", CategoricalArray)`, but I guess the two main uses of `read` for `CategoricalArrays` would be (1) reading in a data frame—in which case, requiring a user to call `categorical!` after reading the data is probably fine for now—(2) reading in a custom struct—in which case, the type will almost certainly have a specified `eltype`. I'm probably make this more complicated than it needs to be.